tla-bmcmt/src/main/scala/at/forsyte/apalache/tla/tracee/pass/TraceePassImpl.scala

```diff
@@ -0,0 +1,116 @@
+        case ex     => exName -> ex
+      }
+    }.toMap
+    require(expressions.nonEmpty)
+
+    /*
+    In this mode, we will inject our own
+    Init and Nexts, but we keep CInit. Since CInit is normally
+    turned into assignment-form within TransitionPass, which is skipped here,
+    we must manually modify and re-insert it.
+     */
+
+    val cinitOpt =
+      derivedPreds.cinit.map { cinitName =>
+        val cinitEx = findBodyOf(cinitName, module.operDeclarations: _*)
+        // We don't perform the standard assignment-search on cinit,
+        // we just replace EVERY x' = e with x' <- e
+        val tr = AssignmentOperatorIntroduction({ _ => true }, tracker)
+        val newEx = tr(cinitEx)
+        ModuleAdapter.exprToOperDef(NormalizedNames.CONST_INIT, newEx)
+      }
+
+    // Trace evaluation uses one variable per provided expression (for which we can reuse the name)
+    val varDecls = expressions.map { case (exName, exVal) =>
+      TlaVarDecl(exName)(exVal.typeTag)
+    }
+
+    // DTC constructs, for each trace state s, the transition that drives the new specification into a state,
+    // where its variables hold the values the expressions would, if evaluated in state s.
+    val dtc = new DrivenTransitionConstructor(tracker, expressions)
+
+    // 1st state on the trace is init, the rest are nexts (possibly empty)
+    val initEx +: nextExs = trace.map(dtc.txToState)
+
+    val initOper = ModuleAdapter.exprToOperDef(NormalizedNames.INIT_PREFIX, initEx)
+    val nextOpers = ModuleAdapter.exprsToOperDefs(NormalizedNames.NEXT_PREFIX, nextExs)
+
+    // We drop certain operators in the original specification, e.g. all expression definitions, as they clash
+    // with the newly-introduced variables.
+    // We also don't need any of the exploration-driving operators.
+    // Notably, we drop the old CInit (it's not in assignment form!), even though we use constants.
+    val droppedOperNames = Set(
+        derivedPreds.init,
+        derivedPreds.next,
+    ) ++ derivedPreds.invariants ++ derivedPreds.cinit ++ expressions.keySet
+
+    /* TODO: Check that no operator has a dependency on the expression-defining operators. If such an operator exists, throw/abort. */
+
+    // Drop old init/next/inv/cinit and expr defs, swap in new init/next/cinit
+    // We keep the rest, since the provided expressions may have operator dependencies
+    val newDecls =
+      varDecls ++ module.constDeclarations ++ module.assumeDeclarations ++
+        module.operDeclarations.filterNot(d => droppedOperNames.contains(d.name)) ++ (initOper +: nextOpers) ++ cinitOpt
+
+    val outModule = TlaModule(module.name, newDecls.toSeq)
+
+    // Mark introduced transition operators as persistent, so they don't get wiped by the inlinig pass.
+    derivedPreds.addPersistent(initOper.name +: nextOpers.map(_.name): _*)
+
+    writeOut(writerFactory, outModule)
+
+    Right(outModule)
```

Comment:
Instead of relying on exception raising for the errors for this, it would help with integration with the RPC system if the error from this is returned as a `Left`.